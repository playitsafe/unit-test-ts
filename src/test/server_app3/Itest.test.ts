import { Account } from "../../app/server_app/model/AuthModel";
import { Reservation } from "../../app/server_app/model/ReservationModel";
import {
  HTTP_CODES,
  HTTP_METHODS,
} from "../../app/server_app/model/ServerModel";
import { Server } from "../../app/server_app/server/Server";
import { makeAwesomeRequest } from "./utils/http-client";
import * as generated from "../../app/server_app/data/IdGenerator";

// Run this test against a working server instance
describe("Sever app integration test suite", () => {
  let server: Server;

  // we are not mocking anything here, we test the real server
  beforeAll(() => {
    server = new Server();
    server.startServer();
  });

  afterAll(() => {
    server.stopServer();
  });

  const someUser: Account = {
    id: "",
    userName: "someUserName",
    password: "somePassword",
  };

  const someReservation: Reservation = {
    id: "",
    endDate: "someEndDate",
    startDate: "someStartDate",
    room: "someRoom",
    user: "someUser",
  };

  //   use ndoe fetch
  it("should register new user", async () => {
    const result = await fetch("http://localhost:8080/register", {
      method: HTTP_METHODS.POST,
      body: JSON.stringify(someUser),
    });

    const resultBody = await result.json();
    expect(result.status).toBe(HTTP_CODES.CREATED);
    expect(resultBody.userId).toBeDefined();
  });

  // use http request module
  it("should register new user with awesomeRequest", async () => {
    const result = await makeAwesomeRequest(
      {
        host: "localhost",
        port: 8080,
        method: HTTP_METHODS.POST,
        path: "/register",
      },
      someUser
    );

    expect(result.statusCode).toBe(HTTP_CODES.CREATED);
    expect(result.body.userId).toBeDefined();
    console.log(`Env var is ${process.env.HOST}`);
  });

  let token: string;
  it("should login a register user", async () => {
    const result = await fetch("http://localhost:8080/login", {
      method: HTTP_METHODS.POST,
      body: JSON.stringify(someUser),
    });
    const resultBody = await result.json();

    expect(result.status).toBe(HTTP_CODES.CREATED);
    expect(resultBody.token).toBeDefined();
    token = resultBody.token;
  });

  let createdReservationId: string;
  it("should create reservation if authorized", async () => {
    const result = await fetch("http://localhost:8080/reservation", {
      method: HTTP_METHODS.POST,
      body: JSON.stringify(someReservation),
      headers: {
        authorization: token,
      },
    });
    const resultBody = await result.json();

    expect(result.status).toBe(HTTP_CODES.CREATED);
    expect(resultBody.reservationId).toBeDefined();
    createdReservationId = resultBody.reservationId;
  });

  it("snapshot test", async () => {
    jest.spyOn(generated, "generateRandomId").mockReturnValueOnce("12345");
    await fetch("http://localhost:8080/reservation", {
      method: HTTP_METHODS.POST,
      body: JSON.stringify(someReservation),
      headers: {
        authorization: token,
      },
    });

    const getResult = await fetch(`http://localhost:8080/reservation/12345`, {
      method: HTTP_METHODS.GET,
      headers: {
        authorization: token,
      },
    });
    const getRequestBody: Reservation = await getResult.json();

    expect(getRequestBody).toMatchSnapshot();
    expect(getRequestBody).toMatchSnapshot();
  });
});
