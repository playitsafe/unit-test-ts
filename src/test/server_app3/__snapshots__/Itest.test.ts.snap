// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sever app integration test suite snapshot test 1`] = `
{
  "endDate": "someEndDate",
  "id": "12345",
  "room": "someRoom",
  "startDate": "someStartDate",
  "user": "someUser",
}
`;

exports[`Sever app integration test suite snapshot test 2`] = `
{
  "endDate": "someEndDate",
  "id": "12345",
  "room": "someRoom",
  "startDate": "someStartDate",
  "user": "someUser",
}
`;
